File: clients/dotnet/src/CanonicalBytes.cs
using System.Buffers;
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Quidnug.Client;

/// <summary>
/// Canonical signable-bytes encoder — byte-for-byte compatible with
/// the Go, Python, Java, Rust, and JavaScript Quidnug SDKs.
///
/// <para>Two entry points:</para>
/// <list type="bullet">
///   <item><see cref="V1Of"/> — v1.0-spec-conformant. Preserves
///         caller-supplied top-level key order (which must mirror
///         Go struct declaration order for the tx type); sorts
///         nested object keys alphabetically to match Go's
///         <c>encoding/json</c> default for
///         <c>map[string]interface{}</c>.</item>
///   <item><see cref="Of"/> — legacy fully-sorted mode. Sorts
///         every level, including the top. Does NOT match the
///         v1.0 canonical form; signatures produced this way
///         will not verify against a v1.0 node. Deprecated.</item>
/// </list>
///
/// <para>Non-ASCII characters emit as raw UTF-8 (not \uXXXX
/// escapes). We write our own JSON serializer rather than using
/// <see cref="JsonSerializer"/> because even
/// <c>JavaScriptEncoder.UnsafeRelaxedJsonEscaping</c> emits
/// supplementary-plane code points as <c>\uXXXX\uXXXX</c>, which
/// would break cross-SDK interop.</para>
/// </summary>
public static class CanonicalBytes
{
    /// <summary>
    /// v1.0-conformant canonical bytes: preserve top-level
    /// caller insertion order, sort nested object keys
    /// alphabetically.
    ///
    /// <para>Pass transaction fields in Go struct declaration
    /// order (the schema for each tx type in
    /// <c>docs/protocol-v1.0.md</c> §4). Any nested
    /// <see cref="JsonObject"/> inside the root will have its
    /// keys sorted before serialization.</para>
    /// </summary>
    public static byte[] V1Of(object obj, params string[] excludeFields)
    {
        if (obj is null) throw new ArgumentNullException(nameof(obj));

        JsonNode? node = obj is JsonNode n
            ? n.DeepClone()
            : JsonSerializer.SerializeToNode(obj);
        if (node is not JsonObject root)
            throw new ArgumentException("expected an object at the root", nameof(obj));

        foreach (string f in excludeFields) root.Remove(f);

        var buf = new ArrayBufferWriter<byte>(256);
        WriteTopLevelObject(buf, root);
        return buf.WrittenSpan.ToArray();
    }

    /// <summary>
    /// Legacy fully-sorted canonical bytes. Sorts every level,
    /// including the top. Does NOT match the v1.0 canonical
    /// form; retained for backward compatibility with pre-v1.0
    /// signing paths.
    /// </summary>
    [Obsolete("Use V1Of for signatures bound for a v1.0 node.")]
    public static byte[] Of(object obj, params string[] excludeFields)
    {
        if (obj is null) throw new ArgumentNullException(nameof(obj));

        JsonNode? node = obj is JsonNode n
            ? n.DeepClone()
            : JsonSerializer.SerializeToNode(obj);
        if (node is not JsonObject root)
            throw new ArgumentException("expected an object at the root", nameof(obj));

        foreach (string f in excludeFields) root.Remove(f);

        var buf = new ArrayBufferWriter<byte>(256);
        WriteNode(buf, root);
        return buf.WrittenSpan.ToArray();
    }

    private static void WriteNode(IBufferWriter<byte> buf, JsonNode? n)
    {
        switch (n)
        {
            case null:
                WriteAscii(buf, "null");
                return;
            case JsonObject obj:
                WriteObject(buf, obj);
                return;
            case JsonArray arr:
                WriteArray(buf, arr);
                return;
            case JsonValue val:
                WriteValue(buf, val);
                return;
        }
    }

    /// <summary>
    /// Write the root object preserving the caller's key order.
    /// Nested objects are written via <see cref="WriteObject"/>,
    /// which sorts their keys (so nested maps match Go's
    /// alphabetical map-marshal output).
    /// </summary>
    private static void WriteTopLevelObject(IBufferWriter<byte> buf, JsonObject obj)
    {
        buf.Write("{"u8);
        bool first = true;
        foreach (var kv in obj)
        {
            if (!first) buf.Write(","u8);
            first = false;
            WriteString(buf, kv.Key);
            buf.Write(":"u8);
            WriteNode(buf, kv.Value);
        }
        buf.Write("}"u8);
    }

    private static void WriteObject(IBufferWriter<byte> buf, JsonObject obj)
    {
        buf.Write("{"u8);
        var keys = obj.Select(kv => kv.Key).OrderBy(k => k, StringComparer.Ordinal).ToList();
        for (int i = 0; i < keys.Count; i++)
        {
            if (i > 0) buf.Write(","u8);
            WriteString(buf, keys[i]);
            buf.Write(":"u8);
            WriteNode(buf, obj[keys[i]]);
        }
        buf.Write("}"u8);
    }

    private static void WriteArray(IBufferWriter<byte> buf, JsonArray arr)
    {
        buf.Write("["u8);
        for (int i = 0; i < arr.Count; i++)
        {
            if (i > 0) buf.Write(","u8);
            WriteNode(buf, arr[i]);
        }
        buf.Write("]"u8);
    }

    private static void WriteValue(IBufferWriter<byte> buf, JsonValue v)
    {
        var elem = v.GetValue<JsonElement>();
        switch (elem.ValueKind)
        {
            case JsonValueKind.String:
                WriteString(buf, elem.GetString() ?? string.Empty);
                break;
            case JsonValueKind.Number:
                WriteNumber(buf, elem);
                break;
            case JsonValueKind.True:
                WriteAscii(buf, "true");
                break;
            case JsonValueKind.False:
                WriteAscii(buf, "false");
                break;
            case JsonValueKind.Null:
                WriteAscii(buf, "null");
                break;
            default:
                // Object/Array shouldn't appear inside JsonValue.
                WriteAscii(buf, "null");
                break;
        }
    }

    private static void WriteNumber(IBufferWriter<byte> buf, JsonElement elem)
    {
        // Try int64 first (most numbers in Quidnug payloads are integer-
        // valued) so we avoid "1.0" when the source was 1. Fall back to
        // decimal / double.
        if (elem.TryGetInt64(out long l))
        {
            WriteAscii(buf, l.ToString(CultureInfo.InvariantCulture));
            return;
        }
        if (elem.TryGetDouble(out double d))
        {
            WriteAscii(buf, d.ToString("R", CultureInfo.InvariantCulture));
            return;
        }
        WriteAscii(buf, elem.GetRawText());
    }

    private static void WriteString(IBufferWriter<byte> buf, string s)
    {
        buf.Write("\""u8);
        // JSON requires escaping: " \ and control chars below 0x20.
        // Non-ASCII characters ARE emitted as raw UTF-8.
        int i = 0;
        while (i < s.Length)
        {
            int code;
            int consumed;
            if (char.IsHighSurrogate(s[i]) && i + 1 < s.Length && char.IsLowSurrogate(s[i + 1]))
            {
                code = char.ConvertToUtf32(s[i], s[i + 1]);
                consumed = 2;
            }
            else
            {
                code = s[i];
                consumed = 1;
            }

            switch (code)
            {
                case '"':  buf.Write("\\\""u8); break;
                case '\\': buf.Write("\\\\"u8); break;
                case '\n': buf.Write("\\n"u8); break;
                case '\r': buf.Write("\\r"u8); break;
                case '\t': buf.Write("\\t"u8); break;
                case '\b': buf.Write("\\b"u8); break;
                case '\f': buf.Write("\\f"u8); break;
                default:
                    if (code < 0x20)
                    {
                        // control char
                        WriteAscii(buf, $"\\u{code:x4}");
                    }
                    else
                    {
                        // emit raw UTF-8
                        string slice = s.Substring(i, consumed);
                        int bytesNeeded = Encoding.UTF8.GetByteCount(slice);
                        Span<byte> span = buf.GetSpan(bytesNeeded);
                        int written = Encoding.UTF8.GetBytes(slice, span);
                        buf.Advance(written);
                    }
                    break;
            }
            i += consumed;
        }
        buf.Write("\""u8);
    }

    private static void WriteAscii(IBufferWriter<byte> buf, string s)
    {
        Span<byte> span = buf.GetSpan(s.Length);
        for (int i = 0; i < s.Length; i++) span[i] = (byte)s[i];
        buf.Advance(s.Length);
    }
}

File: clients/java/src/main/java/com/quidnug/client/CanonicalBytes.java
package com.quidnug.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Canonical signable-bytes encoder — byte-for-byte compatible with the
 * Go, Python, Rust, and JavaScript Quidnug SDKs.
 *
 * <p>This class ships two helpers:
 *
 * <dl>
 *   <dt>{@link #v1Of(Map, String...)} — v1.0-spec-conformant</dt>
 *   <dd>Top-level keys preserve caller insertion order (must match
 *       the Go struct declaration order for the tx type). Nested
 *       {@code Map<String,Object>} payloads and attributes are
 *       recursively sorted alphabetically (matches Go's
 *       {@code encoding/json} default for
 *       {@code map[string]interface{}}).
 *       <br><b>Use this for any transaction bound for a v1.0 node.</b></dd>
 *
 *   <dt>{@link #of(Map, String...)} — legacy fully-sorted mode</dt>
 *   <dd>Sorts <em>every</em> level of the tree alphabetically,
 *       including the top-level keys. This was the pre-v1.0 encoding.
 *       It does <b>not</b> match the v1.0 canonical form and
 *       signatures produced this way will not verify against a
 *       v1.0 node. Kept only for backward compatibility with the
 *       pre-v1.0 {@code QuidnugClient} code path.</dd>
 * </dl>
 *
 * <p>Until {@code QuidnugClient} is fully converted to v1.0 wire
 * shapes (tracked as a separate task), direct users of the Java SDK
 * that need to sign v1.0 transactions should build their own
 * top-level field lists (mirroring
 * {@code docs/test-vectors/v1.0/README.md}) and pass them through
 * {@link #v1Of(Map, String...)}. The {@code VectorsTest} class in
 * {@code src/test/java} has worked examples.
 */
public final class CanonicalBytes {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private CanonicalBytes() {}

    /**
     * v1.0-conformant canonical bytes: preserve top-level caller
     * insertion order, sort nested object keys alphabetically.
     *
     * <p>Pass transaction fields into the {@code obj} map in Go
     * struct declaration order for the tx type. Any nested
     * {@code Map<String,Object>} fields (payload / attributes) will
     * have their keys sorted before serialization.
     */
    public static byte[] v1Of(Map<String, Object> obj, String... excludeFields) {
        try {
            JsonNode tree = MAPPER.valueToTree(obj);
            if (!tree.isObject()) {
                throw new IllegalArgumentException("expected a JSON object at the root");
            }
            ObjectNode shallow = (ObjectNode) tree;
            for (String f : excludeFields) {
                shallow.remove(f);
            }
            JsonNode normalized = sortNestedKeysOnly(shallow);
            return MAPPER.writeValueAsString(normalized).getBytes(StandardCharsets.UTF_8);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("canonicalize: " + e.getMessage(), e);
        }
    }

    /**
     * Legacy fully-sorted canonical bytes. Sorts every level.
     *
     * <p>Does <b>not</b> match the v1.0 canonical form. Kept only
     * for backward compatibility with pre-v1.0 signing paths.
     *
     * @deprecated since 2.0; use {@link #v1Of(Map, String...)} for
     *             signatures bound for a v1.0 node.
     */
    @Deprecated
    public static byte[] of(Map<String, Object> obj, String... excludeFields) {
        try {
            JsonNode tree = MAPPER.valueToTree(obj);
            if (!tree.isObject()) {
                throw new IllegalArgumentException("expected a JSON object at the root");
            }
            ObjectNode shallow = (ObjectNode) tree;
            for (String f : excludeFields) {
                shallow.remove(f);
            }
            JsonNode sorted = sortKeysDeep(shallow);
            return MAPPER.writeValueAsString(sorted).getBytes(StandardCharsets.UTF_8);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("canonicalize: " + e.getMessage(), e);
        }
    }

    /**
     * Recursively sort the keys of every object in the tree — including
     * the root. Used only by the legacy {@link #of(Map, String...)}
     * helper; new code should prefer {@link #sortNestedKeysOnly(JsonNode)}.
     */
    static JsonNode sortKeysDeep(JsonNode n) {
        if (n.isObject()) {
            ObjectNode src = (ObjectNode) n;
            List<String> keys = new ArrayList<>();
            src.fieldNames().forEachRemaining(keys::add);
            Collections.sort(keys);
            ObjectNode out = JsonNodeFactory.instance.objectNode();
            for (String k : keys) {
                out.set(k, sortKeysDeep(src.get(k)));
            }
            return out;
        }
        if (n.isArray()) {
            ArrayNode out = JsonNodeFactory.instance.arrayNode(n.size());
            for (JsonNode item : n) {
                out.add(sortKeysDeep(item));
            }
            return out;
        }
        return n;
    }

    /**
     * Preserve the root object's key order, but recursively sort
     * any nested object's keys alphabetically. Arrays preserve
     * index order; their elements are recursively normalized.
     */
    static JsonNode sortNestedKeysOnly(JsonNode n) {
        if (n.isObject()) {
            ObjectNode src = (ObjectNode) n;
            ObjectNode out = JsonNodeFactory.instance.objectNode();
            src.fieldNames().forEachRemaining(k -> out.set(k, sortKeysDeep(src.get(k))));
            return out;
        }
        if (n.isArray()) {
            ArrayNode out = JsonNodeFactory.instance.arrayNode(n.size());
            for (JsonNode item : n) {
                out.add(sortKeysDeep(item));
            }
            return out;
        }
        return n;
    }
}
